enhancements/microshift/microshift-apiserver-custom-certs.md

```diff
@@ -0,0 +1,188 @@
+  similiar approach is taken by OpenShift.
+
+
+## Design Details
+N/A
+
+### Open Questions [optional]
+- How to handle kubeconfig generation for NamedCerts?
+
+### Test Plan
+add e2e test that will:
+- generate and sign certificates with custom ca.
+- change the default Microshift configuration to use the newly generated certs.
+- make sure system is functional using  generated external kubeconfig.
+
+### Graduation Criteria
+#### Dev Preview -> Tech Preview
+- Gather feedback from early adopters on possible issues.
+
+#### Tech Preview -> GA
+- Extensive testing with various certificates variations.
+
+#### Removing a deprecated feature
+N/A
+
+### Upgrade / Downgrade Strategy
+N/A
+
+### Version Skew Strategy
+N/A
+
+### Operational Aspects of API Extensions
+N/A
+
+#### Failure Modes
+the provided certs value will be validated before is it passed to the api-server flag
+
+this check will prevent Microshift service from starting:
+1. certificates files exists in the disk and readable by microshift process.
+1. certificates shouldnt override the internal local certificates which are managed internally.
+
+this check display warning message at the log and service will be started:
+1. certificates is expired.
+
+
+#### Support Procedures
+configured certs value are prepended to the  `tls-sni-cert-key`  TLS handshake command line flag  which is passed to the kube-apiserver:
+```
+> journalctl -u microshift -b0 | grep tls-sni-cert-key
+
+Jan 24 14:53:00 localhost.localdomain microshift[45313]: kube-apiserver I0124 14:53:00.649099   45313 flags.go:64] FLAG: --tls-sni-cert-key="[/home/eslutsky/dev/certs/server.crt,/home/eslutsky/dev/certs/server.key;/var/lib/microshift/certs/kube-apiserver-external-signer/kube-external-serving/server.crt,/var/lib/microshift/certs/kube-apiserver-external-signer/kube-external-serving/server.key;/var/lib/microshift/certs/kube-apiserver-localhost-signer/kube-apiserver-localhost-serving/server.crt,/var/lib/microshift/certs/kube-apiserver-localhost-signer/kube-apiserver-localhost-serving/server.key;/var/lib/microshift/certs/kube-apiserver-service-network-signer/kube-apiserver-service-network-serving/server.crt,/var/lib/microshift/certs/kube-apiserver-service-network-signer/kube-apiserver-service-network-serving/server.key
+```
+
+Connecting to API server using external SNI yields external certificate.
+   ```shell
+   $ openssl s_client -connect <SNI_ADDRESS>:6443 -showcerts | openssl x509 -text -noout -in - | grep -C 1 "Alternative\|CN"
+   ```
+
+## Implementation History
+N/A
+
+## Alternatives
+- Use cert-manager for managing Microshift external custom certs
+  - Benifits
+    1. will allow Microshift Administrator create automatic Certificate rotation strategy using supported protocols  (ie: ACME Webhook)
+
+  - Drawbacks
+    1. cert-manager is for managing certs inside k8s clusters. (its not designed for kube-api certs which is based on files.)
```

Comment:
This is inaccurate. One of cert-manager's documented use-cases is to provide access to apiserver using 3rd party certs: https://docs.google.com/document/d/10eC-AWFRnzdCzkbYNDBOvlObYdjG2jiCr4wUKVrEGqE/edit

It's immaterial whether or not the certs provided to secure apiserver's primary endpoint are provided as files, as it does not affect other means of exposing the endpoint as documented in the aforementioned documentation.

Reply:
it does require that the API server will be managed by  [kube-apiserver](https://github.com/openshift/cluster-kube-apiserver-operator)   , which currently is not the case of Microshift. it sounds like a very big architecture change .